### PR TITLE
code cleanup

### PR DESCRIPTION
**Pull request by izulin**
### Context

General code cleanup. Prevents some bugs that were *possible* (but not encountered yet), simplifies code, unifies usage of `null` vs `undefined`.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1.
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [ ] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

File: src/interpreter/plugin/LookupPlugin.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ import {ProcedureAst} from '../../parser'
 import {StatType} from '../../statistics'
 import {zeroIfEmpty} from '../ArithmeticHelper'
 import {InterpreterState} from '../InterpreterState'
-import {InternalScalarValue, InterpreterValue, RawNoErrorScalarValue, } from '../InterpreterValue'
+import {InternalScalarValue, InterpreterValue, RawNoErrorScalarValue,} from '../InterpreterValue'
```
Comment (wojciechczerniak) [inline src/interpreter/plugin/LookupPlugin.ts:15]: Linter reported problem here
```suggestion
import {InternalScalarValue, InterpreterValue, RawNoErrorScalarValue} from '../InterpreterValue'
```

File: src/Exporter.ts [hunk 2]
```diff
@@ -4,11 +4,11 @@
  */
 
 import {CellError, ErrorType, SimpleCellAddress, simpleCellAddress} from './Cell'
-import {CellValue, DetailedCellError, } from './CellValue'
+import {CellValue, DetailedCellError,} from './CellValue'
 import {Config} from './Config'
 import {CellValueChange, ChangeExporter} from './ContentChanges'
 import {ErrorMessage} from './error-message'
-import {EmptyValue, getRawValue, InterpreterValue, isExtendedNumber, } from './interpreter/InterpreterValue'
+import {EmptyValue, getRawValue, InterpreterValue, isExtendedNumber,} from './interpreter/InterpreterValue'
```
Comment (wojciechczerniak) [inline src/Exporter.ts:11]: Linter
```suggestion
import {EmptyValue, getRawValue, InterpreterValue, isExtendedNumber} from './interpreter/InterpreterValue'
```

File: src/Exporter.ts [hunk 3]
```diff
@@ -4,11 +4,11 @@
  */
 
 import {CellError, ErrorType, SimpleCellAddress, simpleCellAddress} from './Cell'
-import {CellValue, DetailedCellError, } from './CellValue'
+import {CellValue, DetailedCellError,} from './CellValue'
```
Comment (wojciechczerniak) [inline src/Exporter.ts:7]:
```suggestion
import {CellValue, DetailedCellError} from './CellValue'
```
Linter

File: src/DependencyGraph/RangeVertex.ts [hunk 4]
```diff
@@ -70,10 +70,8 @@ export class RangeVertex {
    * @param cacheKey - key to retrieve from the cache
    * @param criterionString - criterion text (ex. '<=5')
    */
-  public getCriterionFunctionValue(cacheKey: string, criterionString: string): any | null {
-    const values = this.getCriterionFunctionValues(cacheKey)
-    const value = values.get(criterionString)
-    return value ? value[0] : null
+  public getCriterionFunctionValue(cacheKey: string, criterionString: string) {
```
Comment (wojciechczerniak) [inline src/DependencyGraph/RangeVertex.ts:73]: Type disapeard
```suggestion
  public getCriterionFunctionValue(cacheKey: string, criterionString: string): ?? {
```

File: src/DependencyGraph/FormulaCellVertex.ts [hunk 5]
```diff
@@ -241,10 +241,9 @@ export class FormulaCellVertex extends FormulaVertex {
     version: number,
   ) {
     super(formula, address, version)
-    this.cachedCellValue = null
   }
 
-  public valueOrNull(): InterpreterValue | null {
+  public valueOrNull(): Maybe<InterpreterValue> {
```
Comment (wojciechczerniak) [inline src/DependencyGraph/FormulaCellVertex.ts:246]: Method name is `valueOrNull` but it's `undefined` now. Does it make sense or it doesn't matter?

File: src/DependencyGraph/FormulaCellVertex.ts [hunk 6]
```diff
@@ -116,9 +116,9 @@ export class MatrixVertex extends FormulaVertex {
     return this.matrix.simpleRangeValue()
   }
 
-  public valueOrNull(): InterpreterValue | null {
+  public valueOrNull(): Maybe<InterpreterValue> {
```
Comment (wojciechczerniak) [inline src/DependencyGraph/FormulaCellVertex.ts:119]: same `valueOrNull` returns `undefined`

File: src/DependencyGraph/FormulaCellVertex.ts [hunk 7]
```diff
@@ -73,7 +73,7 @@ export abstract class FormulaVertex {
    */
   public abstract getCellValue(): InterpreterValue
 
-  public abstract valueOrNull(): InterpreterValue | null
+  public abstract valueOrNull(): Maybe<InterpreterValue>
```
Comment (wojciechczerniak) [inline src/DependencyGraph/FormulaCellVertex.ts:76]: and here

File: src/interpreter/plugin/NumericAggregationPlugin.ts [hunk 8]
```diff
@@ -12,7 +12,7 @@ import {Ast, AstNodeType, CellRangeAst, ProcedureAst} from '../../parser'
 import {ColumnRangeAst, RowRangeAst} from '../../parser/Ast'
 import {coerceBooleanToNumber} from '../ArithmeticHelper'
 import {InterpreterState} from '../InterpreterState'
-import {EmptyValue, ExtendedNumber, getRawValue, InternalScalarValue, isExtendedNumber, } from '../InterpreterValue'
+import {EmptyValue, ExtendedNumber, getRawValue, InternalScalarValue, isExtendedNumber,} from '../InterpreterValue'
```
Comment (wojciechczerniak) [inline src/interpreter/plugin/NumericAggregationPlugin.ts:15]:
```suggestion
import {EmptyValue, ExtendedNumber, getRawValue, InternalScalarValue, isExtendedNumber} from '../InterpreterValue'
```
Linter

File: src/parser/addressRepresentationConverters.ts [hunk 9]
```diff
@@ -166,7 +178,7 @@ export function columnIndexToLabel(column: number) {
   return result.toUpperCase()
 }
 
-export function sheetIndexToString(sheetId: number, sheetMappingFn: SheetIndexMappingFn): string | undefined {
+export function sheetIndexToString(sheetId: number, sheetMappingFn: SheetIndexMappingFn): Maybe<string> {
```
Comment (wojciechczerniak) [inline src/parser/addressRepresentationConverters.ts:181]: in some places you've changed `Maybe<string>` into `string | undefined` but here the opposite?

File: src/parser/addressRepresentationConverters.ts [hunk 10]
```diff
@@ -31,54 +31,66 @@ export const cellAddressFromString = (sheetMapping: SheetMappingFn, stringAddres
...
-  const sheet = extractSheetNumber(result, sheetMapping)
+  let sheet = extractSheetNumber(result, sheetMapping)
   if (sheet === undefined) {
     return undefined
   }
 
+  if(sheet === null) {
+    sheet = undefined
+  }
+
```
Comment (wojciechczerniak) [inline src/parser/addressRepresentationConverters.ts:87]: When `sheet` is `undefined` we exit early, but then we assign `undefined` and process it. Shouldn't we exit early for both `null` and `undefined`?

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
Looks nice and clean overall 👍

File: src/parser/addressRepresentationConverters.ts (hunk 10 above)
Comment (izulin) [reply src/parser/addressRepresentationConverters.ts:87]: Yeah, this part is a mess. Believe me, the logic should be as it is (there are two distinct failure modes here), and it is pretty annoying. I really don't see any clean way of making it less ugly, though.

File: src/DependencyGraph/RangeVertex.ts (hunk 4 above)
Comment (izulin) [reply src/DependencyGraph/RangeVertex.ts:73]: the type was `any | null` so basically `any`, which is not really informative (and triggers linter...)

File: src/parser/addressRepresentationConverters.ts (hunk 9 above)
Comment (izulin) [reply src/parser/addressRepresentationConverters.ts:181]: I changed to `T | undefined` in the `HyperFormula.ts` so the API does not introduce extra object.
Internally I prefer we use `Maybe<T>` wherever it makes sense.

Comment (wojciechczerniak) [reply src/parser/addressRepresentationConverters.ts:181]: Sounds legit 👍

File: src/DependencyGraph/RangeVertex.ts (hunk 4 above)
Comment (wojciechczerniak) [reply src/DependencyGraph/RangeVertex.ts:73]: I see. Better without type then

**Review by wojciechczerniak — APPROVED** [review]

**Pull request merged**